### PR TITLE
[FLINK-14301][doc] add documentation for functions categories and new function resolution orders

### PR DESCRIPTION
**Pull request by bowenli86**
### What is the purpose of the change

- unifies function related docs in a single dir/tab
- add general introduction of function categories, reference mechanism, and resolution orders

## Brief change log

- renamed functions.md to builtinFunctions.md
- unifies function related docs in a single dir/tab of docs/dev/table/functions/
- add general introduction of function categories, reference mechanism, and resolution orders

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

**Comment by bowenli86** [comment]
@sjwiesman @xuefuz @lirui-apache @zjuwangg

**Comment by bowenli86** [comment]
docs of how to register and manipulate such functions will be added later, as ddl and apis are still moving around and not finalized yet

**Comment by dawidwys** [comment]
@flinkbot attention @fhueske @sjwiesman

File: docs/dev/table/functions/index.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+-->
+
+Flink Table API & SQL empowers users to do data transformations with functions.
+
+Types of Functions
+------------------
+
+There are two dimensions to classify functions in Flink.
+
+One dimension is system (or built-in) functions v.s. catalog functions. System functions have no namespace and can be
```
Comment (sjwiesman) [inline docs/dev/table/functions/index.md]: You say system here but named the page built in. Let's just pick one and stick with it.

File: docs/dev/table/functions/index.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+ lifespan of a session, they are always created by users. Persistent functions live across lifespan of sessions, they are either
+ provided by the system or persisted in catalogs.
+ 
+The two dimensions give Flink users 4 categories of functions:
+
+1. Temporary system functions
+2. System functions
+3. Temporary catalog functions
+4. Catalog functions
+
```
Comment (sjwiesman) [inline docs/dev/table/functions/index.md]: This is conflating properties that I don't feel are related. built-in vs catalog and temporary vs persistent should be seperate sections.

File: docs/dev/table/functions/index.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+The other dimension is temporary functions v.s. persistent functions. Temporary functions are volatile and only live up to
+ lifespan of a session, they are always created by users. Persistent functions live across lifespan of sessions, they are either
+ provided by the system or persisted in catalogs.
+ 
+The two dimensions give Flink users 4 categories of functions:
+
+1. Temporary system functions
+2. System functions
+3. Temporary catalog functions
+4. Catalog functions
```
Comment (sjwiesman) [inline docs/dev/table/functions/index.md]: I think this is unnecessary and can be removed.

File: docs/dev/table/functions/index.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+Flink Table API & SQL empowers users to do data transformations with functions.
+
```
Comment (sjwiesman) [inline docs/dev/table/functions/index.md]: Add toc

File: docs/dev/table/functions/index.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+Precise function reference empowers users to use catalog functions specifically, and across catalog and across database, 
+e.g. `select mycatalog.mydb.myfunc(x) from mytable` and `select mydb.myfunc(x) from mytable`.
+
+This is only supported starting from Flink 1.10.
+
+## Ambiguous Function Reference
+
+In ambiguous function reference, users just specify the function's name in SQL query, e.g. `select myfunc(x) from mytable`.
+
+This is the only supported option before Flink 1.10.
```
Comment (sjwiesman) [inline docs/dev/table/functions/index.md]: Remove, if a user is running an old version of Flink they can reference that versions docs.

File: docs/dev/table/functions/index.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+The resolution order is:
+
+1. Temporary catalog function
+2. Catalog function
+
+## Ambiguous Function Reference
+
+In Flink 1.10, we also redefined its function resolution order to be:
+
+The new order is:
```
Comment (sjwiesman) [inline docs/dev/table/functions/index.md]: No need to qualify,  if a user is running an old version of Flink they can reference that versions docs.

File: docs/dev/table/functions/index.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
+---
+title: "Functions"
+nav-id: table_functions
+nav-parent_id: tableapi
+nav-pos: 116
```
Comment (sjwiesman) [inline docs/dev/table/functions/index.md]: This renders very low down, I think functions should be higher up on the side nav.

File: docs/dev/table/functions/index.md (hunk 2 above)
Comment (bowenli86) [reply docs/dev/table/functions/index.md]: I actually feel they are related to help users understand resolution order below. Maybe I should have mentioned that in resolution order, system functions always precede catalog's and temporary functions always precede persistent on its own dimension

File: docs/dev/table/functions/index.md (hunk 3 above)
Comment (bowenli86) [reply docs/dev/table/functions/index.md]: ditto

File: docs/dev/table/functions/index.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+is_beta: true
```
[the remark this answers is not in the document]
Comment (bowenli86) [reply docs/dev/table/functions/index.md]: removed

File: docs/dev/table/functions/index.md (hunk 1 above)
Comment (bowenli86) [reply docs/dev/table/functions/index.md]: let's use "system"

File: docs/dev/table/functions/index.md (hunk 6 above)
Comment (bowenli86) [reply docs/dev/table/functions/index.md]: the old version actually doesn't talk about this. and this will be a disruptive change

**Comment by bowenli86** [comment]
@sjwiesman thanks for your review. pls take another look

File: docs/dev/table/functions/index.md (hunk 6 above)
Comment (sjwiesman) [reply docs/dev/table/functions/index.md]: I don't understand this change well enough to be able to respond to that, however, I generally feel these kinds of changes are what release notes are for. I'd like to hear @dawidwys or @fhueske opinion and if they say it stays then +1 from me.

**Comment by bowenli86** [comment]
@flinkbot attention @sjwiesman @dawidwys @fhueske

File: docs/dev/table/functions/index.md (hunk 6 above)
Comment (dawidwys) [reply docs/dev/table/functions/index.md]: I agree with @sjwiesman. Personally I think the documentation should just describe the current state. Unless this is a disruptive change that requires users to change their code. I don't think this is the case. I

f I am wrong, and it does, I would mention the version, but also would mention what are the actions that a user must take and would change the title to something along the lines: "Migration guide/Migration"

**Comment by dawidwys** [comment]
Sorry I missed the notifications. I have no major concerns regarding the content. However I would prefer somebody better with the linguistic aspect gave the final approval.

**Comment by bowenli86** [comment]
@sjwiesman I removed the 1.10 info, please take another look

**Comment by sjwiesman** [comment]
+1 to merge

**Pull request closed without merging**